### PR TITLE
do not index internal transactions with simple token transfers

### PR DESCRIPTION
**Pull request by ayrat555**
fixes https://github.com/poanetwork/blockscout/issues/1331

## Motivation

0xa9059cbb - signature of the transfer(address,uint256) function from the ERC-20 token specification.
Although transaction input data can be faked we use these heuristics to filter simple token transfer internal transactions from indexing because they slow down real-time fetcher

## Changelog

- do not index internal transactions with simple token transfers

### CODE REVIEW

File: apps/indexer/lib/indexer/block/realtime/fetcher.ex [hunk 1]
```diff
@@ -332,7 +333,8 @@ defmodule Indexer.Block.Realtime.Fetcher do
        ) do
     case transactions_params
          |> transactions_params_to_fetch_internal_transactions_params(token_transfers_params)
-         |> EthereumJSONRPC.fetch_internal_transactions(json_rpc_named_arguments) do
+         |> EthereumJSONRPC.fetch_internal_transactions(json_rpc_named_arguments)
+         |> reject_simple_token_transfer_internal_transactions() do
```
Comment (KronicDeth) [inline apps/indexer/lib/indexer/block/realtime/fetcher.ex]: This is still fetching internal transactions and only rejecting the output internal transactions, but #1341 is saying that it is the fetching of internal transactions is slow, not the inserting of the internal transactions into the database.

File: apps/indexer/lib/indexer/block/realtime/fetcher.ex [hunk 2]
```diff
@@ -373,6 +375,21 @@ defmodule Indexer.Block.Realtime.Fetcher do
...
   end
 
+  defp reject_simple_token_transfer_internal_transactions({:ok, internal_transaction_params}) do
+    # 0xa9059cbb - signature of the transfer(address,uint256) function from the ERC-20 token specification.
+    # Although transaction input data can be faked we use this heuristics to filter simple token transfer internal transactions from indexing because they slow down realtime fetcher
+
+    filtered_internal_transaction =
+      Enum.reject(internal_transaction_params, fn internal_transaction ->
+        internal_transaction.value == 0 &&
+          String.starts_with?(internal_transaction.input, TokenTransfer.transfer_function_signature())
```
Comment (zachdaniel) [inline apps/indexer/lib/indexer/block/realtime/fetcher.ex]: We should check here to determine if the input data can actually be decoded as the function `transfer(address, uint256)`. That would avoid false negatives. For instance, someone could cause a collision, but the function that collided to get the same results could be `some_other_random_function(uint256, uint256, uint256)`. We'd only catch that it couldn't match if we *also* try to decode the input accordingly.

File: apps/indexer/lib/indexer/block/realtime/fetcher.ex (hunk 1 above)
Comment (ayrat555) [reply apps/indexer/lib/indexer/block/realtime/fetcher.ex]: @KronicDeth I don't think there is a way to check if a transaction is token transfer without sending a request to ethereum node

Comment (ayrat555) [reply apps/indexer/lib/indexer/block/realtime/fetcher.ex]: @KronicDeth I stopped fetching internal transactions from simple token transfer transactions

File: apps/indexer/lib/indexer/block/realtime/fetcher.ex (hunk 2 above)
Comment (ayrat555) [reply apps/indexer/lib/indexer/block/realtime/fetcher.ex]: @zachdaniel I addressed your comment. And now input parameters are being decoded

**Pull request merged**